### PR TITLE
Describe adding insert support in a connector

### PR DESCRIPTION
**Pull request by nineinchnick**
## Description

> Is this change a fix, improvement, new feature, refactoring, or other?
improvement

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)
docs

> How would you describe this change to a non-technical end user or system administrator?
n/a

## Related issues, pull requests, and links

* ``ConnectorPageSinkProvider`` is also briefly described in #12956, it could link to this page or maybe since it's short we should merge these two PRs. OTOH we might expand this more by adding examples on how to iterate over a `Page`, handle transactions, delayed writes, etc.

## Documentation

( ) No documentation is needed.
(x) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

(x) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: docs/src/main/sphinx/develop/insert.rst [hunk 1]
```diff
@@ -0,0 +1,30 @@
+=============================================
+Supporting ``INSERT`` and ``CREATE TABLE AS``
+=============================================
+
+To support ``INSERT``, a connector must implement:
+
+* ``beginInsert()`` and ``finishInsert`` from the ``ConnectorMetadata``
```
Comment (findinpath) [inline docs/src/main/sphinx/develop/insert.rst:7]: `finishInsert` -> `finishInsert()`

File: docs/src/main/sphinx/develop/insert.rst [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  interface;
+* a ``ConnectorPageSinkProvider`` that receives a table handle and returns
+  a ``ConnectorPageSink``.
+
+When executing an ``INSERT`` statement, the engine calls the ``beginInsert()``
+method in the connector, which receives a table handle and a list of column.
```
Comment (findinpath) [inline docs/src/main/sphinx/develop/insert.rst:13]: `column` -> `columns`

File: docs/src/main/sphinx/develop/insert.rst [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+It should return a ``ConnectorInsertTableHandle``, that can carry any
+connector specific information and will be passed to the page sink provider.
+The page sink provider creates a page sink, that should accept ``Page`` objects.
```
Comment (findinpath) [inline docs/src/main/sphinx/develop/insert.rst:16]: _page sink provider_  --> `PageSinkProvider`

File: docs/src/main/sphinx/develop/insert.rst [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+* ``beginInsert()`` and ``finishInsert()`` from the ``ConnectorMetadata``
+  interface;
+* a ``ConnectorPageSinkProvider`` that receives a table handle and returns
+  a ``ConnectorPageSink``.
+
+When executing an ``INSERT`` statement, the engine calls the ``beginInsert()``
+method in the connector, which receives a table handle and a list of columns.
+It should return a ``ConnectorInsertTableHandle``, that can carry any
+connector specific information and will be passed to the page sink provider.
```
Comment (mosabua) [inline docs/src/main/sphinx/develop/insert.rst:15]:
```suggestion
connector specific information, and it is passed to the page sink provider.
```

File: docs/src/main/sphinx/develop/insert.rst [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+The ``PageSinkProvider`` creates a page sink, that should accept ``Page`` objects.
```
Comment (mosabua) [inline docs/src/main/sphinx/develop/insert.rst:16]:
```suggestion
The ``PageSinkProvider`` creates a page sink, that accepts ``Page`` objects.
```

File: docs/src/main/sphinx/develop/insert.rst [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+When all the pages for a specific split have been processed, the Trino engine
```
Comment (mosabua) [inline docs/src/main/sphinx/develop/insert.rst:18]:
```suggestion
When all the pages for a specific split have been processed, Trino calls
```

File: docs/src/main/sphinx/develop/insert.rst [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+calls ``ConnectorPageSink.finish()``, which returns a ``Collection<Slice>``
+of "fragments" representing connector-specific information about
```
Comment (mosabua) [inline docs/src/main/sphinx/develop/insert.rst:20]: drop the ""

File: docs/src/main/sphinx/develop/insert.rst [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+the processed rows.
+
+When all pages for all splits have been processed, the Trino engine calls
```
Comment (mosabua) [inline docs/src/main/sphinx/develop/insert.rst:23]: Just use "Trino" in general .. dont talk about "the Trino engine"

File: docs/src/main/sphinx/develop/insert.rst [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+``ConnectorMetadata.finishInsert()``, passing a collection containing all
+the "fragments" from all the splits.  The connector does what is required
+to finalize the operation, for example, committing the transaction.
+
+To support ``CREATE TABLE AS``, the ``ConnectorPageSinkProvider`` should
```
Comment (mosabua) [inline docs/src/main/sphinx/develop/insert.rst:28]: must also

**Review by mosabua — CHANGES_REQUESTED** [review]

File: docs/src/main/sphinx/develop/insert.rst (hunk 4 above)
Comment (nineinchnick) [reply docs/src/main/sphinx/develop/insert.rst:15]: I forgot about using the present tense, but I wanted to ask about passive voice - is `it is passed` passive and should we change it to say that the engine passes it?

File: docs/src/main/sphinx/develop/insert.rst (hunk 7 above)
Comment (nineinchnick) [reply docs/src/main/sphinx/develop/insert.rst]: I copied this from the existing `delete and update` section. I fixed it there too, in a separate commit.

**Comment by nineinchnick** [comment]
@mosabua ready for another round, I also ran the Vale linter on this and it found one extra space :D

File: docs/src/main/sphinx/develop/delete-and-update.rst [hunk 10]
```diff
@@ -24,11 +24,11 @@ methods to get a "rowId" column handle; to start the operation; and to finish th
   ``UpdatablePageSource`` ``deleteRows()`` or ``updateRows()`` method.  Those methods persist
   the deletions or updates in the underlying data store.
 * When all the pages for a specific split have been processed, the Trino engine calls
-  ``UpdatablePageSource.finish()``, which returns a ``Collection<Slice>`` of "fragments"
+  ``UpdatablePageSource.finish()``, which returns a ``Collection<Slice>`` of fragments
   representing connector-specific information about the rows processed by the calls to
   ``deleteRows`` or ``updateRows``.
 * When all pages for all splits have been processed, the Trino engine calls ``ConnectorMetadata.finishDelete()`` or
-  ``finishUpdate``, passing a collection containing all the "fragments" from all the splits.  The connector
+  ``finishUpdate``, passing a collection containing all the fragments from all the splits.  The connector
```
Comment (mosabua) [inline docs/src/main/sphinx/develop/delete-and-update.rst:31]: can we remove the double space and wrap this?

**Review by mosabua — APPROVED** [review]
One nit but essentially good to go.

File: docs/src/main/sphinx/develop/delete-and-update.rst (hunk 10 above)
Comment (nineinchnick) [reply docs/src/main/sphinx/develop/delete-and-update.rst]: That opened a small can of worms. I ran the Vale linter on it and found some other minor typos.

File: docs/src/main/sphinx/develop/delete-and-update.rst [hunk 11]
```diff
@@ -16,28 +16,28 @@ methods to get a "rowId" column handle; to start the operation; and to finish th
   connector's ``ConnectorPageSource``, to read pages on behalf of the Trino engine, and to
   write deletions and/or updates to the underlying data store.
 * The connector's ``UpdatablePageSource.getNextPage()`` implementation fetches the next page
-  from the underlying ``ConnectorPageSource``, optionally reformats the page, and returns it
+  from the underlying ``ConnectorPageSource``, optionally rebuild the page, and returns it
```
Comment (martint) [inline docs/src/main/sphinx/develop/delete-and-update.rst:19]: > rebuilds

File: docs/src/main/sphinx/develop/insert.rst [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+To support ``INSERT``, a connector must implement:
+
+* ``beginInsert()`` and ``finishInsert()`` from the ``ConnectorMetadata``
+  interface;
+* a ``ConnectorPageSinkProvider`` that receives a table handle and returns
+  a ``ConnectorPageSink``.
+
+When executing an ``INSERT`` statement, the engine calls the ``beginInsert()``
+method in the connector, which receives a table handle and a list of columns.
+It should return a ``ConnectorInsertTableHandle``, that can carry any
```
Comment (martint) [inline docs/src/main/sphinx/develop/insert.rst:14]: The grammar of this sentence is off. cc @mosabua

**Pull request merged**